source/custom/nimxlcterminaladaptor_restricted_service.custom.cpp

```diff
@@ -0,0 +1,80 @@
+#include <iostream>
+#include <vector>
+#include <server/converters.h>
+#include <nierr_Status.h>
+#include <nimxlcterminaladaptor_restricted.pb.h>
+#include <nimxlcterminaladaptor_restricted/nimxlcterminaladaptor_restricted_service.h>
+#include <custom/nimxlcterminaladaptor_restricted_converters.h>
+
+namespace nimxlcterminaladaptor_restricted_grpc {
+  
+  using nidevice_grpc::converters::allocate_output_storage;
+  using nidevice_grpc::converters::convert_to_grpc;
+  using nimxlcterminaladaptor_restricted_grpc::NIErrStatus;
+  
+::grpc::Status NimxlcTerminalAdaptorRestrictedService::GetDeviceContainer(::grpc::ServerContext* context, const GetDeviceContainerRequest* request, GetDeviceContainerResponse* response)
+{
+    if (context->IsCancelled()) {
+        return ::grpc::Status::CANCELLED;
+    }
+    try {
+        auto grpc_session = request->session();
+        nimxlc_Session session = session_repository_->access_session(grpc_session.id(), grpc_session.name());
+
+        nierr_Status status;
+        status.code = 0;
+        status.capacity = 0;
+        status.json = "";
```

Comment:
`status->json` (`status.json` here) is normally allocated on the heap using `status->reallocJson`. Initializing `status->json` to NULL is ok but initializing it to a string literal is risky. If capacity > 0, nimxlc uses code in `nierr_Status.c` that overwrites the buffer. This would corrupt the EXE's string constant table (or probably crash because string constants are usually read-only).

When nimxlc sets a fatal status, it uses code in `nierr_Status.c` that calls `status->reallocJson`. If you leave `reallocJson` uninitialized, setting a fatal status will crash.

Before the function returns, it should use `status->reallocJson` to free `status->json`. Otherwise, it will leak memory. Most clients of `nierr_Status` use the C++ `nierr::Status` class to do this.

I think it would be possible to use a fixed-size, stack-allocated buffer if you initialize `status->capacity` to the buffer size and `status->reallocJson` to point to a function that always returns false.

Cc: @irwand

Reply:
Why are we initializing status manually like that? why not use the provided `nierr_Status_initialize(&status)` ?

Reply:
or.. yeah, just use the `nierr::Status` C++ wrapper as suggested.

Reply:
I think this is the bare minimum fix for now:
- Initialize `status.reallocJson` to point to a function with the correct signature that returns false.
- Initialize `status.json` to `nullptr`
- Update the rest of this code to tolerate a NULL `status.json`

It won't return any JSON, but at least it shouldn't crash.